File: Week 1/13.How Integer is Stored.java
/*
 * WEEK 1 - JAVA FUNDAMENTALS
 * Topic: How Integers are Stored in Memory
 * File: 13.How Integer is Stored.java
 *
 * CONCEPT:
 * Explains the binary representation of integers in Java. Positive numbers
 * are stored as their direct binary equivalent, while negative numbers use
 * Two's Complement representation. This is fundamental to understanding
 * integer overflow and bitwise operations.
 *
 * KEY POINTS:
 * - int is a signed 32-bit type (range: -2,147,483,648 to 2,147,483,647)
 * - Positive integers: directly converted to binary
 * - Negative integers: stored as Two's Complement
 * - Two's Complement steps: (1) get binary of absolute value,
 *   (2) flip all bits (1's complement), (3) add 1
 * - Two's Complement has only ONE representation for zero (unlike sign-magnitude)
 * - Subtraction can be performed using addition logic with Two's Complement
 *
 * Space Complexity: int always uses exactly 4 bytes (32 bits) regardless of value
 */

/*
a) How are integers stored ?


The most commonly used integer type is int which is a signed 32-bit type. When you store an integer, its corresponding binary value is stored. The way
integers are stored differs for negative and positive numbers. For positive numbers the integral value is simple converted into binary value and for negative
numbers their 2’s compliment form is stored.


Let’s discuss How are Negative Numbers Stored?


Computers use 2's complement in representing signed integers because:
1. There is only one representation for the number zero in 2's complement, instead of two representations in sign-magnitude and 1's complement.
2. Positive and negative integers can be treated together in addition and subtraction. Subtraction can be carried out using the "addition logic".


Example:   */

int i = -4;

/*
Steps to calculate Two’s Complement of -4 are as follows:

Step 1: Take Binary Equivalent of the positive value (4 in this case)
0000 0000 0000 0000 0000 0000 0000 0100


Step 2: Write 1's complement of the binary representation by inverting the bits
1111 1111 1111 1111 1111 1111 1111 1011


Step 3: Find 2's complement by adding 1 to the corresponding 1's complement
1111 1111 1111 1111 1111 1111 1111 1011
+0000 0000 0000 0000 0000 0000 0000 0001
------------------------------------------------------------
1111 1111 1111 1111 1111 1111 1111 1100


Thus, integer -4 is represented by the binary sequence (1111 1111 1111 1111
1111 1111 1111 1100) in Java.
*/

File: Week 1/cpp/13.how_integer_is_stored.cpp
/*
 * WEEK 1 - C++ FUNDAMENTALS
 * Topic: How Integers are Stored in Memory
 * File: 13.how_integer_is_stored.cpp
 *
 * CONCEPT:
 * Like Java, C++ stores signed integers in two's-complement form on virtually
 * every modern platform (since C++20 the standard formally REQUIRES it).
 * Negative numbers reuse the same hardware addition logic as positives.
 *
 * KEY POINTS:
 *  - sizeof(int) is 4 bytes on most desktops, but the standard only guarantees >= 2.
 *  - For positive n: bit pattern == binary of n.
 *  - For negative n: bit pattern == 2^N + n (mod 2^N).
 *  - Two's complement: invert the bits of |n|, then add 1.
 *  - Bitwise operators (& | ^ ~ << >>) work directly on the underlying bits.
 *  - Use std::bitset<32> for nice binary printing.
 *
 * SYNTAX:
 *   #include <bitset>
 *   std::cout << std::bitset<32>(n) << '\n';
 *
 * DRY RUN — derive the 32-bit pattern of -4:
 *  1. +4  = 0000_0000_0000_0000_0000_0000_0000_0100
 *  2. ~4  = 1111_1111_1111_1111_1111_1111_1111_1011
 *  3. +1  = 1111_1111_1111_1111_1111_1111_1111_1100  (== -4 in two's complement)
 */

#include <iostream>
#include <bitset>
#include <cstdint>
#include <climits>

int main() {
    std::cout << "sizeof(int)  = " << sizeof(int)  << " bytes ("
              << sizeof(int) * 8 << " bits)\n";
    std::cout << "INT_MIN      = " << INT_MIN << '\n';
    std::cout << "INT_MAX      = " << INT_MAX << '\n';

    int positive = 4;
    int negative = -4;

    std::cout << "\n=== +4 ===\n";
    std::cout << "decimal: " << positive << '\n';
    std::cout << "32-bit : " << std::bitset<32>(positive) << '\n';

    std::cout << "\n=== -4 (two's complement) ===\n";
    std::cout << "decimal: " << negative << '\n';
    std::cout << "32-bit : " << std::bitset<32>(negative) << '\n';

    // Manual derivation
    std::int32_t step1 = 4;                          // +|n|
    std::int32_t step2 = ~step1;                      // ones complement
    std::int32_t step3 = step2 + 1;                   // twos complement
    std::cout << "\nstep 1) +4 binary       = " << std::bitset<32>(step1) << '\n';
    std::cout << "step 2) inverted bits   = " << std::bitset<32>(step2) << '\n';
    std::cout << "step 3) plus 1 (== -4) = " << std::bitset<32>(step3) << '\n';

    // Overflow demo (signed wraparound is technically UB; unsigned is well-defined)
    std::uint32_t u = 0xFFFFFFFFu;
    std::cout << "\nunsigned 0xFFFFFFFF + 1 = " << (u + 1) << "  (wraps to 0)\n";

    return 0;
}

/*
 * NOTES:
 *  - C++ guarantees two's complement only since C++20; earlier it was
 *    implementation-defined (in practice every mainstream compiler used 2C).
 *  - Signed overflow is UB in C++; unsigned overflow wraps modulo 2^N.
 *  - <bitset> provides cheap stringified binary printing.
 *  - For arbitrary-precision integers, consider Boost.Multiprecision.
 */
